plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/RegisterTableProcedure.java

```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    private final TrinoCatalogFactory catalogFactory;
+    private final ClassLoader classLoader;
+
+    @Inject
+    public RegisterTableProcedure(TrinoCatalogFactory catalogFactory)
+    {
+        this.catalogFactory = requireNonNull(catalogFactory, "catalogFactory is null");
+        // this class is loaded by PluginClassLoader, and we need its reference to be stored
+        this.classLoader = getClass().getClassLoader();
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                SYSTEM_SCHEMA,
+                PROCEDURE_NAME,
+                ImmutableList.of(
+                        new Procedure.Argument(SCHEMA_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_LOCATION, VARCHAR),
+                        new Procedure.Argument(METADATA_FILE_NAME, VARCHAR, false, null)),
+                REGISTER_TABLE.bindTo(this));
+    }
+
+    public void registerTable(ConnectorSession clientSession, String schemaName, String tableName, String tableLocation, String metadataFileName)
+    {
+        // this line guarantees that classLoader that we stored in the field will be used inside try/catch
+        // as we captured reference to PluginClassLoader during initialization of this class
+        // we can use it now to correctly execute the procedure
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+            doRegisterTable(
+                    clientSession,
+                    schemaName,
+                    tableName,
+                    tableLocation,
+                    Optional.ofNullable(metadataFileName));
+        }
+    }
+
+    private void doRegisterTable(ConnectorSession clientSession, String schemaName, String tableName, String tableLocation, Optional<String> metadataFileName)
+    {
+        if (schemaName == null || tableName == null || tableLocation == null) {
+            throw new TrinoException(INVALID_ARGUMENTS, "Illegal parameter set passed. " + PROCEDURE_USAGE_EXAMPLES);
+        }
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+        TrinoCatalog catalog = catalogFactory.create(clientSession.getIdentity());
+        if (!catalog.namespaceExists(clientSession, schemaTableName.getSchemaName())) {
+            throw new TrinoException(SCHEMA_NOT_FOUND, format("Schema '%s' does not exist", schemaTableName.getSchemaName()));
+        }
+        if (catalog.tableExists(clientSession, schemaTableName)) {
+            throw new TrinoException(TABLE_ALREADY_EXISTS, format("Table '%s' already exists in schema '%s'", schemaTableName.getTableName(), schemaTableName.getSchemaName()));
+        }
+        String metadataLocation = metadataFileName
+                .map(name -> format("%s/%s/%s", stripTrailingSlash(tableLocation), METADATA_FOLDER_NAME, name))
```

Comment:
Are we restricting user to provide the metadata file name in the format of `format("%05d-%s%s", newVersion, randomUUID(), getFileExtension(codec)`? If yes then we can validate the metadata file name, Else `validateLocation` will be sufficient to validate the fully qualified metadata file location. I will be moving `validateLocation` check in `RegisterTableProcedure`.

https://github.com/trinodb/trino/pull/14375/files/434e5d8f98368f7ad3a9aafacb7f7c73f95bebbb#diff-d3f0d86c9a9f041e92e1db2d851a1623d7ab60473f8d3bcc0bf5aedcf0bab829:~:text=public%20static%20void-,validateLocation,-(TrinoFileSystem%20fileSystem

Reply:
> Are we restricting user to provide the metadata file name in the format of `format("%05d-%s%s", newVersion, randomUUID(), getFileExtension(codec)`?

no, we should allow arbitrary file name here.



> Else `validateLocation` will be sufficient to validate the fully qualified metadata file location.

yes, except that it's not our intent to support metadata files living somewhere else (in a different folder)

on a file system like HDFS, one could try provide `../../../../some-metadata.json` and it could work, while we don't want to support this use-case, so we want to reject it cleanly.

Reply:
TableMetadataParser does not allow reading metadata files unless the file name contains `.metadata.json`. 
https://github.com/apache/iceberg/blob/cd68f9cbad8e7a6048096e8531883bfad3399c8d/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L275

https://github.com/apache/iceberg/blob/cd68f9cbad8e7a6048096e8531883bfad3399c8d/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L69-L71

I'll put the validation for metadata file name.

1. should not be empty (If provided as an argument)
2. should contain `.metadata.json`
3. should not contain any prefixed or suffixed slash 

Anything missing?

Reply:
> 2\. should contain `.metadata.json`

i'd leave it up to the library

Reply:
Done

Reply:
When `metadala_file_name` is an existing file But not a valid metadata file (like `.avro` file in `metadata` folder), this still registers the table.

However `select` query fails with `java.lang.IllegalArgumentException: hdfs://hadoop-master:9000/user/hive/warehouse/test-1dff2cfc382a43c88017a87abaacd4c5/metadata/01fc4fb9-2f1c-49d5-8500-7928c8cd92f0-m0.avro is not a valid metadata file`. 

Should we validate the metadata content before registering the table?

CC: @alexjo2144 | @findinpath | @findepi

Reply:
Makes sense to me to validate the metadata file.

Of course, it doesn't guarantee table is readable (problems like data files may be corrupted or missing; the table can have been copied with metadata pointing to the old location, etc). However, especially given metadata file name can be provided by the user, it makes sense to validate that.